### PR TITLE
Fix threshold over rate metric resulting in `NaN` result

### PR DESCRIPTION
**Pull request by oleiade**
This Pull Request attempts to address #2520. It was discovered internally by the team that thresholds applied to checks (rate metrics) would immediately fail, and return a `NaN` result.

I've found out that the `NaN` was the result of a division by zero. I assumed it happens because the threshold was run before a single check sample was recorded (I might be wrong there though). I've thus introduced a check in `Thresholds.Run` to perform insert the sink's "rate" value computation only if the right-hand side of the division is not zero. 

Besides, the `Threshold.runNoTaint` function used to make a verification that a sink existed, before actually evaluating the threshold expression. This PR drops this verification, considering the thresholds are now verified at startup. Because of the previously mentioned issue, it would have led to a situation where we know the metric exists, we know the threshold is valid for this metric, but we fail with an error because at a certain point of the execution, our internal storage doesn't have data for it yet. 

I'm not a 100% confident this PR doesn't have unintended side-effects. I have tried it which a bunch of scripts of mine who try to exhibit as much of possible thresholds use cases, and none seem to be broken by this change. But I'd take any more insights if you have. The point of a hotfix being, after all, to fix stuff, not potentially introduce more problems 🤞🏻 

**Learning point**: good to know that `int64(0) / int64(0)` leads to a panic, but `float64(0) / float64(0)` leads to `NaN`.

### CODE REVIEW

**Review by na-- — DISMISSED** [review]
Hmm I am not sure I fully grok what's going on here. 

From what I understand, this problem didn't appear before, in k6 v0.37.0 and below, because we basically only started running the threshold calculations for a metric once k6 saw a sample for the metric, is that right? And this PR kind of reverts that behavior, for `Rate` metrics at least? 

Because `0/0` is NaN, and we can't even pretend it's `0`, since that will be invalid as well? So it's best to just not run the threshold until there is some data. 

If I've understood things correctly, LGTM, though I don't know some of the internals here very well now.

**Review by codebien — APPROVED** [review]
The code seems fine to me

>Learning point: good to know that int64(0) / int64(0) leads to a panic, but float64(0) / float64(0) leads to NaN.

This is strange, they should be the same result, right? Is a type wrapped by `goja`? :thinking:

**Comment by oleiade** [comment]
@codebien it's actually a behavior you can observe in pure Go: https://go.dev/play/p/BVaocakjKXD

**Comment by oleiade** [comment]
@mstoykov opinions, and ideas to improve https://github.com/grafana/k6/pull/2521/commits/0116c5d82aba978378d4c36c50036a20e3b0dc5f ?

File: core/engine.go [hunk 1]
```diff
@@ -299,7 +311,7 @@ func (e *Engine) processMetrics(globalCtx context.Context, processMetricsAfterRu
 			if !e.runtimeOptions.NoThresholds.Bool {
 				// Ensure the ingester flushes any buffered metrics
 				_ = e.ingester.Stop()
-				thresholdsTainted, _ := e.MetricsEngine.EvaluateThresholds()
+				thresholdsTainted, _ := e.MetricsEngine.EvaluateThresholds(true)
```
Comment (mstoykov) [inline core/engine.go:314]: I think you continued my mistake with the boolean logic :rofl:. 

We don't should ignore empty sinks everywhere except here. Also my `IsEmpty` code is all wrong ... so you basically need to reverse to logic (or check for `== 0` instead of `>0` ).

 ```suggestion
				thresholdsTainted, _ := e.MetricsEngine.EvaluateThresholds(false)
```

File: core/engine.go [hunk 2]
```diff
@@ -226,7 +231,7 @@ func (e *Engine) startBackgroundProcesses(
 			for {
 				select {
 				case <-ticker.C:
-					thresholdsTainted, shouldAbort := e.MetricsEngine.EvaluateThresholds()
+					thresholdsTainted, shouldAbort := e.MetricsEngine.EvaluateThresholds(false)
```
Comment (mstoykov) [inline core/engine.go:234]:
```suggestion
					thresholdsTainted, shouldAbort := e.MetricsEngine.EvaluateThresholds(true)
```

**Review by mstoykov — CHANGES_REQUESTED** [review]
I think you continued my faulty boolean logic :sweat_smile: 

I've left what I think needs to change in order for the code to be correct. 

It will also be nice if we have tests for this as currently there are none :(.

**Comment by oleiade** [comment]
@mstoykov After some more head-scratching and trying to figure out why the engine tests were failing, I think I've found out the issue. My boolean logic was actually correct, I just inverted the one you did in your initial patch. Your logic was additive: `Evalutate` or `EvaluateAll`, and I made it subtractive `Evaluate(` (all is the default), or `Evaluate(without)`. I realize now that I'd rather follow the positive/additive logic though, as it's the main path, and is simpler to understand. Going to modify the flag accordingly 👍🏻

**Comment by oleiade** [comment]
I intend to add integration tests to cover the scenarios which triggered this specific issue separately in #2525.

**Comment by oleiade** [comment]
@mstoykov could I get a thumbs up please? 🙇🏻

**Review by mstoykov — APPROVED** [review]
